### PR TITLE
FEAT: Adding conn.setencoding() API

### PR DESCRIPTION
**Pull request by jahnvi480**
### Work Item / Issue Reference  

> [AB#34917](https://sqlclientdrivers.visualstudio.com/c6d89619-62de-46a0-8b46-70b92a84d85e/_workitems/edit/34917)

-------------------------------------------------------------------
### Summary   
This pull request adds support for configurable text encoding in the `Connection` class of the `mssql_python` package, allowing users to specify how text is encoded when interacting with the database. It introduces new methods for setting and retrieving encoding settings, provides robust validation and error handling, and includes comprehensive tests to ensure correct behavior across a wide range of scenarios.

**Encoding support and configuration:**

* Added `setencoding` and `getencoding` methods to the `Connection` class, allowing users to configure and retrieve text encoding and C type for SQL statements and parameters. The default encoding is set to `'utf-16le'` with `SQL_WCHAR`, and robust validation is performed for both encoding and ctype values.
* Improved error handling in `setencoding`, raising `ProgrammingError` for invalid encodings or ctypes, and `InterfaceError` if the connection is closed. 

**Testing and validation:**

* Added a comprehensive suite of tests for the new encoding functionality, including default settings, explicit and automatic ctype detection, invalid parameter handling, persistence across cursors, thread safety, and case sensitivity. Also includes a (skipped) test for Unicode data operations.

**Documentation and compatibility:**

* Updated docstrings and comments to clarify encoding behavior and DB-API 2.0 compatibility.

These changes make it easier and safer for users to control how text data is encoded and sent to the database, with clear error reporting and extensive test coverage to ensure reliability.

### CODE REVIEW

File: mssql_python/connection.py [hunk 1]
```diff
@@ -12,12 +12,49 @@
...
+# UTF-16 encoding variants that should use SQL_WCHAR by default
+UTF16_ENCODINGS = frozenset([
+    'utf-16',
+    'utf-16le', 
+    'utf-16be'
+])
+
+# Cache for encoding validation to improve performance
+# Using a simple dict instead of lru_cache for module-level caching
+_ENCODING_VALIDATION_CACHE = {}
```
Comment (sumitmsft) [inline mssql_python/connection.py]: The _ENCODING_VALIDATION_CACHE dictionary and _CACHE_MAX_SIZE variable are defined but are not used anywhere in the code. Since _validate_encoding now uses @lru_cache for caching, these variables appear redundant. Consider removing them to keep the codebase clean and avoid confusion for future maintainers.

File: mssql_python/connection.py [hunk 2]
```diff
@@ -12,12 +12,49 @@
...
+_CACHE_MAX_SIZE = 100  # Limit cache size to prevent memory bloat
+
+
+@lru_cache(maxsize=128)
```
Comment (sumitmsft) [inline mssql_python/connection.py]: The cache size in @lru_cache(maxsize=128) is good for most use cases. If the API is used in long-running processes with hundreds of different custom encodings, then we can consider eviction/logging. However, I don't think for a driver, such a situation may arise. We can skip it for now.

File: mssql_python/connection.py [hunk 3]
```diff
@@ -159,6 +204,106 @@ def setautocommit(self, value: bool = False) -> None:
...
+        if not _validate_encoding(encoding):
+            # Log the sanitized encoding for security
+            log('warning', "Invalid encoding attempted: %s", sanitize_user_input(str(encoding)))
+            raise ProgrammingError(
+                driver_error=f"Unsupported encoding: {encoding}",
+                ddbc_error=f"The encoding '{encoding}' is not supported by Python",
+            )
+        
+        # Normalize encoding to lowercase for consistency
+        encoding = encoding.lower()
```
Comment (sumitmsft) [inline mssql_python/connection.py]: Can you think of edge-case Unicode characters? If yes, then for Unicode normalization, we can consider using .casefold().  However, for encoding names, .lower() is generally sufficient and widely used. But, only from the Unicode perspective lower may not be sufficient.

**Review by sumitmsft — CHANGES_REQUESTED** [review]
Left a few comments...

**Review by sumitmsft — DISMISSED** [review]

**Review by gargsaumya — DISMISSED** [review]

**Review by gargsaumya — APPROVED** [review]

**Review by bewithgaurav — APPROVED** [review]
